Error when truncating a table that is externally referenced from another table by specifying the table name.

Thank you for the great tool.
We use it all the time.

When I try to truncate a table that is externally referenced from another table by specifying the table name, I seem to get an invalid memory address or nil pointer dereference.

We think this is the relevant area.
https://github.com/cloudspannerecosystem/spanner-truncate/blob/main/truncate/coordinator.go#L116-L141

If a foreign key reference is set up with a column name different from the table name, It seems that even though the tableMap is registered with schema.tableName, it is referenced by tableMap[referencing] when it is referenced, resulting in a nil reference and a segmentation fault.

For example, this occurs when there is another table that references the "User" table with the column name "ReferenceUserId".

It looks like the only way to get around this is to use the referencing column name in the form tableName+Id.
Is there any other way around this?

@yfuruyama 

Prepare data for a table schema like this.
```
CREATE TABLE Users (
  UserId STRING(36) NOT NULL,
  AccountId STRING(36),
  UpdatedAt TIMESTAMP NOT NULL,
  CreatedAt TIMESTAMP NOT NULL,
) PRIMARY KEY(UserId);

CREATE TABLE Test (
    TestId STRING(36) NOT NULL,
    TestName STRING(255) NOT NULL,
    Object STRING(255) NOT NULL,
    CreatedAt TIMESTAMP NOT NULL,
    UpdatedAt TIMESTAMP NOT NULL
) PRIMARY KEY(TestId);

CREATE TABLE UserTest (
	TestId STRING(36) NOT NULL,
	TesterUserId STRING(36) NOT NULL,
	CONSTRAINT FKTesterUserId FOREIGN KEY (TesterUserId) REFERENCES Users (UserId)
) PRIMARY KEY(TestId, TesterUserId),
INTERLEAVE IN PARENT Test ON DELETE CASCADE;
```


The DML looks like this.
```
INSERT INTO Users (UserId, AccountId, UpdatedAt, CreatedAt) VALUES
('a1b2-c3d4-e5f6-g7h8-i9j0', 'acc1-1234-5678-9101-1121', CURRENT_TIMESTAMP, CURRENT_TIMESTAMP);

INSERT INTO Test (TestId, TestName, Object, CreatedAt, UpdatedAt) VALUES
('1a2b3c4d-5e6f-7g8h-9i0j-1k2l3m4n5o6p', 'Test1', 'Object1', CURRENT_TIMESTAMP, CURRENT_TIMESTAMP);

INSERT INTO UserTest (TestId, TesterUserId) VALUES
('1a2b3c4d-5e6f-7g8h-9i0j-1k2l3m4n5o6p', 'a1b2-c3d4-e5f6-g7h8-i9j0');
```

We used it as such a library in our unit testing with DB.
```
err := truncate.Run(
		ctx,
		{{ project_id }},
		{{ instance_name }},
		{{ database_name }},
		true,
		new(bytes.Buffer),
		[]string{"Users"}, // <---
		[]string{},
	)
```
Make Users the target of truncate, delete the data, then put the data into Users and run the test.
The Test and UserTest tables do nothing.

Then it seems to result in this error.
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x60 pc=0xfd5429]

goroutine 1026 [running]:
github.com/cloudspannerecosystem/spanner-truncate/truncate.(*table).isDeletable(0xc00013cf50)
	/go/pkg/mod/github.com/cloudspannerecosystem/spanner-truncate@v0.1.0/truncate/coordinator.go:54 +0xc9
github.com/cloudspannerecosystem/spanner-truncate/truncate.findDeletableTables({0xc0005ec600, 0x1, 0xc0001aa000?})
	/go/pkg/mod/github.com/cloudspannerecosystem/spanner-truncate@v0.1.0/truncate/coordinator.go:92 +0x8d
github.com/cloudspannerecosystem/spanner-truncate/truncate.(*coordinator).start.func1()
	/go/pkg/mod/github.com/cloudspannerecosystem/spanner-truncate@v0.1.0/truncate/coordinator.go:184 +0x16a
created by github.com/cloudspannerecosystem/spanner-truncate/truncate.(*coordinator).start
	/go/pkg/mod/github.com/cloudspannerecosystem/spanner-truncate@v0.1.0/truncate/coordinator.go:175 +0x95
```

It seems to happen only when the table names are specified in the eighth argument targetTables of truncate.Run.
You can avoid this by specifying an empty slice and executing on all tables.

--- a/truncate/coordinator.go
+++ b/truncate/coordinator.go
@@ -19,6 +19,7 @@ package truncate
 import (
 	"context"
 	"errors"
+	"fmt"
 	"time"
 
 	"cloud.google.com/go/spanner"
@@ -110,7 +111,7 @@ type coordinator struct {
 	errChan chan error
 }
 
-func newCoordinator(schemas []*tableSchema, indexes []*indexSchema, client *spanner.Client) *coordinator {
+func newCoordinator(schemas []*tableSchema, indexes []*indexSchema, client *spanner.Client) (*coordinator, error) {
 	var tables []*table
 	tableMap := map[string]*table{}
 	for _, schema := range schemas {
@@ -136,6 +137,9 @@ func newCoordinator(schemas []*tableSchema, indexes []*indexSchema, client *span
 
 		table := tableMap[schema.tableName]
 		for _, referencing := range schema.referencedBy {
+			if _, ok := tableMap[referencing]; !ok {
+				return nil, fmt.Errorf("%s is referenced by %s, but %s is not in the table list", schema.tableName, referencing, referencing)
+			}
 			table.referencedBy = append(table.referencedBy, tableMap[referencing])
 		}
 	}
@@ -167,7 +171,7 @@ func newCoordinator(schemas []*tableSchema, indexes []*indexSchema, client *span
 	return &coordinator{
 		tables:  topLevelTables,
 		errChan: make(chan error),
-	}
+	}, nil
 }
 
 // start starts coordination in another goroutine.

--- a/truncate/coordinator_test.go
+++ b/truncate/coordinator_test.go
@@ -28,6 +28,7 @@ func TestNewCoordinator(t *testing.T) {
 		schemas []*tableSchema
 		indexes []*indexSchema
 		want    []*table
+		wantErr bool
 	}{
 		{
 			desc: "Flat",
@@ -96,6 +97,14 @@ func TestNewCoordinator(t *testing.T) {
 			},
 		},
 		{
+			desc: "Foreign Key referencing table not exist",
+			schemas: []*tableSchema{
+				{tableName: "A", parentTableName: ""},
+				{tableName: "B", parentTableName: "", referencedBy: []string{"C"}},
+			},
+			wantErr: true,
+		},
+		{
 			desc: "Child table has an interleaved index",
 			schemas: []*tableSchema{
 				{tableName: "A", parentTableName: ""},
@@ -123,7 +132,14 @@ func TestNewCoordinator(t *testing.T) {
 		},
 	} {
 		t.Run(test.desc, func(t *testing.T) {
-			coordinator := newCoordinator(test.schemas, test.indexes, nil)
+			coordinator, err := newCoordinator(test.schemas, test.indexes, nil)
+			if test.wantErr {
+				if err == nil {
+					t.Errorf("test wants error, but no error returned")
+				}
+				return
+			}
+
 			got := coordinator.tables
 			if !compareTables(got, test.want) {
 				t.Errorf("invalid tables: got = %#v, want = %#v", got, test.want)

--- a/truncate/run.go
+++ b/truncate/run.go
@@ -67,7 +67,10 @@ func Run(ctx context.Context, projectID, instanceID, databaseID string, quiet bo
 		return fmt.Errorf("failed to fetch index schema: %v", err)
 	}
 
-	coordinator := newCoordinator(schemas, indexes, client)
+	coordinator, err := newCoordinator(schemas, indexes, client)
+	if err != nil {
+		return fmt.Errorf("failed to coordinate: %v", err)
+	}
 	coordinator.start(ctx)
 
 	// Show progress bars.

